include/merlin/core_kernels/lookup_kernels.cuh

```diff
@@ -0,0 +1,722 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http:///www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "kernel_utils.cuh"
+
+namespace nv {
+namespace merlin {
+
+// Using 32 threads to deal with one key
+template <typename K = uint64_t, typename V = float, typename S = uint64_t,
+          typename CopyScore = CopyScoreEmpty<S, K, 128>,
+          typename CopyValue = CopyValueTwoGroup<float, float4, 32>>
+__global__ void lookup_kernel_with_io_pipeline_v1(
+    const Bucket<K, V, S>* buckets, const size_t buckets_num, const int dim,
+    const K* __restrict keys, V* __restrict values, S* __restrict scores,
+    bool* __restrict founds, size_t n) {
+  constexpr int GROUP_SIZE = 32;
+  constexpr int RESERVE = 16;
+  constexpr int DIM_BUF = 224;
+  constexpr int BLOCK_SIZE = 128;
+  constexpr int BUCKET_SIZE = 128;
```

Comment:
@jiashuy Are they configurable? How do you decide their values?

Reply:
At present:
1. `BUCKET_SIZE`  is fixed to 128. This is commonly used by users which can be confirmed by @rhdong .
2. I think `BLOCK_SIZE` is as small as possible to reduce uneven workload. But too small will cause the grid size too large. So I choose 128.
3. `GROUP_SIZE` is set accorrding the profiler. When dim is small, use 16 threads to deal with one key cooperatively is more effective(if use 8, will consume more registers); and when dim is large, use 32 threads to deal with one key, so that we can put larger value to shared memory(group num is smaller, means using less shared memory for double buffer). 
And the only difference between kernel v1 and v2 is the `GROUP_SIZE`.
4. `DIM_BUF` is configurable, according to the shared memory size of SM(different on arch). I've already finished this, and will commit today.
5. `RESERVE` is the reserved size for possible keys(digest = target digest<some bits of hashed key>).
 From the statistics of continues keys, 16 is enough for `RESERVE`, but I use 8 in `lookup_kernel_with_io_pipeline_v2`，for reduce shared memory usage. Resolving correctness by swaping space with time(latency).
 The frequency of the reserve size that is really needed is a power-law distribution.

Reply:
In a summry, `BUCKET_SIZE`，`BLOCK_SIZE`、`GROUP_SIZE` are fixed for specific kernel. And `BLOCK_SIZE` is set accorrding the subjective experience。
`DIM_BUF` is configurable and have been implemented.
`RESERVE` and `GROUP_SIZE` are set by summary from profiler and performance.